Adding inlay hints in the VS Code extension.

lahodaj opened the pull request:
Adding ability to show the inline hints (using "inlayHints" of the LSP), showing the same hints as for NetBeans. Enabled by default, but can be adjusted either in global settings (Editor/Inlay Hints), or individual Java-specific capabilities can be enabled/disabled in the extension settings.

It looks like this:
![inlay-hints](https://github.com/apache/netbeans/assets/51319204/6477004f-9173-44d6-8fb5-4e08e383cbcc)

---
**^Add meaningful description above**

By opening a pull request you confirm that, unless explicitly stated otherwise, the changes -

 - are all your own work, and you have the right to contribute them.
 - are contributed solely under the terms and conditions of the Apache License 2.0 (see section 5 of the license for more information).

Please make sure (eg. `git log`) that all commits have a valid name and email address for you in the Author field.

If you're a first time contributor, see the Contributing guidelines for more information.

If you're a committer, please label the PR before pressing "Create pull request" so that the right test jobs can run.

jtulach commented on java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/TextDocumentServiceImpl.java:
```diff
@@ -2402,6 +2411,69 @@ protected CallHierarchyOutgoingCall createResultItem(CallHierarchyItem item, Lis
+                    js.runUserActionTask(cc -> {
+                        cc.toPhase(JavaSource.Phase.RESOLVED);
+                        Preferences preferences = NbPreferences.root().node("/org/netbeans/modules/java/editor/InlineHints/default");
+                        preferences.putBoolean(JAVA_INLINE_HINT_PARAMETER_NAME, enabled.contains("parameter"));
```
`Preferences.put`!? Quite a hacky solution. I'd replace this with "branding API"_ and specified different default in `nbcode` suite.

jtulach commented:
From end user perspective, the solution is really great:

![TreeToIr.java](https://github.com/apache/netbeans/assets/1842422/3ee28feb-4c66-49ed-ab52-e9a7276ac42f)

it will make it way easier for me to work with Enso code base.

lahodaj commented:
I've updated the patch, make a general provider (so that other languages can contribute as well), the settings in `package.json` are better now, changed the default in the extension to not produce any inline hints by default. The default in the server if there's no configuration at all is to produce all the hints.

Any insights/feedback? Might be good to finalize this at some point.

Thanks!

jtulach commented on ide/api.lsp/src/org/netbeans/spi/lsp/InlayHintsProvider.java:
```diff
@@ -0,0 +1,74 @@
+@MimeLocation(subfolderName = "CodeLensProvider")
+public interface InlayHintsProvider {
+
+    public Set<String> supportedHintTypes();
```
Missing documentation and explanation of what this is for.

jtulach commented on ide/api.lsp/src/org/netbeans/spi/lsp/InlayHintsProvider.java:
```diff
@@ -0,0 +1,74 @@
+     */
+    public CompletableFuture<List<? extends InlayHint>> codeLens(@NonNull Context context);
+
+    public static final class Context {
```
No documentation.

jtulach commented on ide/api.lsp/src/org/netbeans/spi/lsp/InlayHintsProvider.java:
```diff
@@ -0,0 +1,74 @@
+        private final @NonNull Set<String> requestedHintTypes;
+        private final @NullAllowed Range range;
+
+        public Context(@NonNull FileObject file, @NonNull Set<String> requestedHintTypes, @NullAllowed Range range) {
```
This should be "module private", right?

jtulach commented on java/java.editor.base/src/org/netbeans/modules/java/editor/base/semantic/SemanticHighlighterBase.java:
```diff
@@ -1197,5 +1168,35 @@ public static interface ErrorDescriptionSetter {
+        public static Settings getDefault() {
+            Preferences preferences = NbPreferences.root().node("/org/netbeans/modules/java/editor/InlineHints/default");
+            return new Settings(preferences.getBoolean(JAVA_INLINE_HINT_PARAMETER_NAME, DEFAULT_VALUES.get(JAVA_INLINE_HINT_PARAMETER_NAME)),
+                                preferences.getBoolean(JAVA_INLINE_HINT_CHAINED_TYPES, DEFAULT_VALUES.get(JAVA_INLINE_HINT_CHAINED_TYPES)),
```
How is this connected with `vscode/package.json`?

jtulach reviewed (approved):
Thanks for making this work!

lahodaj commented:
Commit squashed, tests are passing. Unless there are objections, I probably should integrate in the next day or so. @dbalek, @MartinBalin
